src/args/arg_matches.rs

```diff
@@ -1,282 +1,572 @@
+    /// This method returns the `ArgMatches` for a particular subcommand or None if the subcommand
+    /// wasn't present at runtime.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
     /// # use clap::{App, Arg, SubCommand};
-    /// # let app_matches = App::new("myapp")
-    /// #     .subcommand(SubCommand::with_name("test")).get_matches();
-    /// if let Some(matches) = app_matches.subcommand_matches("test") {
-    ///     // Use matches as normal
+    /// let app_m = App::new("myprog")
+    ///     .arg(Arg::with_name("debug")
+    ///         .short("d"))
+    ///     .subcommand(SubCommand::with_name("test")
+    ///         .arg(Arg::with_name("opt")
+    ///             .long("option")
+    ///             .takes_value(true)))
+    ///     .get_matches_from(vec![
+    ///         "myprog", "-d", "test", "--option", "val"
+    ///     ]);
+    ///
+    /// // Both parent commands, and child subcommands can have arguments present at the same times
+    /// assert!(app_m.is_present("debug"));
+    ///
+    /// // Get the subcommand's ArgMatches instance
+    /// if let Some(sub_m) = app_m.subcommand_matches("test") {
+    ///     // Use the struct like normal
+    ///     assert_eq!(sub_m.value_of("opt"), Some("val"));
     /// }
     /// ```
-    pub fn subcommand_matches<'na>(&self, name: &'na str) -> Option<&ArgMatches> {
-        if let Some(ref sc) = self.subcommand {
-            if sc.name != name {
-                return None;
-            }
-            return Some(&sc.matches);
-        }
-        None
+    pub fn subcommand_matches<S: AsRef<str>>(&self, name: S) -> Option<&ArgMatches<'a>> {
+        self.subcommand.as_ref().map(|s| if s.name == name.as_ref() { Some(&s.matches) } else { None } ).unwrap()
     }
 
-    /// Returns the name of the subcommand used of the parent `App`, or `None` if one wasn't found
+    /// Because subcommands are essentially "sub-apps" they have their own `ArgMatches` as well.
+    /// But simply getting the sub-`ArgMatches` doesn't help much if we don't also know which
+    /// subcommand was actually used. This method returns the name of the subcommand that was used
+    /// at runtime, or `None` if one wasn't.
     ///
-    /// *NOTE*: Only a single subcommand may be present per `App` at runtime, does *NOT* check for
-    /// the name of sub-subcommand's names
+    /// *NOTE*: Subcommands form a hierarchy, where multiple subcommands can be used at runtime,
+    /// but only a single subcommand from any group of sibling commands may used at once.
     ///
+    /// An ASCII art depiction may help explain this better...Using a fictional version of `git` as
+    /// the demo subject. Imagine the following are all subcommands of `git` (note, the author is
+    /// aware these aren't actually all subcommands in the real `git` interface, but it makes
+    /// explaination easier)
+    ///
+    /// ```notrust
```

Comment:
This is required due to a bug in `rustdoc`. Otherwise we get a panic, even when using `ignore`. See [this Rust issue](https://github.com/rust-lang/rust/issues/12834)